mllib/src/test/scala/org/apache/spark/ml/util/MLTestingUtils.scala

```diff
@@ -224,4 +208,139 @@ object MLTestingUtils extends SparkFunSuite {
+    }
+  }
+
+  /**
+   * Helper function for testing sample weights. Tests that oversampling each point is equivalent
+   * to assigning a sample weight proportional to the number of samples for each point.
+   */
+  def testOversamplingVsWeighting[M <: Model[M], E <: Estimator[M]](
+        spark: SparkSession,
+        estimator: E with HasWeightCol with HasLabelCol with HasFeaturesCol,
+        categoricalFeaturesInfo: Map[Int, Int],
+        numPoints: Int,
+        numClasses: Int,
+        numFeatures: Int,
+        modelEquals: (M, M) => Unit,
+        seed: Long): Unit = {
+    import spark.implicits._
+    val df = generateNoisyData(numPoints, numClasses, numFeatures, categoricalFeaturesInfo,
+      seed).toDF()
+    val (overSampledData, weightedData) = genEquivalentOversampledAndWeightedInstances(
+      df, estimator.getLabelCol, estimator.getFeaturesCol, seed)
+    val weightedModel = estimator.set(estimator.weightCol, "weight").fit(weightedData)
+    val overSampledModel = estimator.set(estimator.weightCol, "").fit(overSampledData)
+    modelEquals(weightedModel, overSampledModel)
+  }
+
+  /**
+   * Helper function for testing sample weights. Tests that injecting a large number of outliers
+   * with very small sample weights does not affect fitting. The predictor should learn the true
+   * model despite the outliers.
+   */
+  def testOutliersWithSmallWeights[M <: Model[M], E <: Estimator[M]](
+        spark: SparkSession,
+        estimator: E with HasWeightCol with HasLabelCol with HasFeaturesCol,
+        categoricalFeaturesInfo: Map[Int, Int],
+        numPoints: Int,
+        numClasses: Int,
+        numFeatures: Int,
+        modelEquals: (M, M) => Unit,
+        seed: Long): Unit = {
+    import spark.implicits._
+    val df = generateNoisyData(numPoints, numClasses, numFeatures, categoricalFeaturesInfo,
+      seed).toDF()
+    val outlierFunction = getRandomLinearPredictionFunction(numFeatures, numClasses, seed - 1)
+    val outlierDF = df.as[Instance].flatMap { case Instance(l, w, f) =>
+      List.fill(3)(Instance(outlierFunction(f), 0.0001, f)) ++ List(Instance(l, w, f))
+    }
+    val trueModel = estimator.set(estimator.weightCol, "").fit(df)
+    val outlierModel = estimator.set(estimator.weightCol, "weight").fit(outlierDF)
+    modelEquals(trueModel, outlierModel)
+  }
+
+  /**
+   * Helper function for testing sample weights. Tests that giving constant weights to each data
+   * point yields the same model, regardless of the magnitude of the weight.
+   */
+  def testArbitrarilyScaledWeights[M <: Model[M], E <: Estimator[M]](
+      data: Dataset[LabeledPoint],
+      estimator: E with HasWeightCol with HasLabelCol with HasFeaturesCol,
+      modelEquals: (M, M) => Unit): Unit = {
+    estimator
+      .set(estimator.labelCol, "label")
+      .set(estimator.featuresCol, "features")
+      .set(estimator.weightCol, "weight")
+    val models = Seq(0.001, 1.0, 1000.0).map { w =>
```

Comment:
I disagree. 1.0 and 1000.0 are both integers, and I have already had experience with algorithms not properly handling fractional weights before. I think here we cover the case of (tiny, unit, and large) weights as well as (fractional, integer) weights.